basic implementation of volumes

nupfel opened the pull request.

bartekrutkowski commented:
Thanks for the submision! Do you mind providing full usage/testing commands along with their output?

nupfel commented:
sorry it took so long to get it all ready to be merged!

can you give me an example of what you mean by "full usage/testing commands"? i added documentation about usage with example commands in `doc/source/advanced-use.rst` as well as the manpage.

bartekrutkowski commented on doc/source/advanced-use.rst:
```diff
@@ -249,3 +249,30 @@ Now simply create a jail and supply the pkgs.txt file:
+
+        /source/path(:/destination/path(:mount_options)) /another/source/path...
+
+``source_path`` is the minimum required input and causes ``destination_path`` to be the same as ``source_path``, but located within the jail's root path.
```
I think 'parameter' works better than 'input' here.

bartekrutkowski commented on doc/source/advanced-use.rst:
```diff
@@ -249,3 +249,30 @@ Now simply create a jail and supply the pkgs.txt file:
+
+Volumes can be added at creation time:
+
+``iocell create volumes="/path/to/some/host/dir:/destination/within/jail/root /other/host/dir:/dest/in/jail:ro" tag=myjail``
```
I think the example path using 'host' world is a bit misleading here, suggesting cross-host, over the network, functionality.

rspeed replied:
"source" seems like it would be the correct term.

rostwald replied:
I think "jailhost" is used for this kind of reference in most documentations.

nupfel replied:
if you search for "host" in any of the documentation you will find it has always been used to refer to the jail host, not another host across the network, RTFM.
i can see how this may be a little more confusing in this case using volumes, hence why i changed it in the upcoming fix, but "host" it is definitely more consistent. i'm starting to question the usefulness of this fork to be honest. i doubt it will ever become mainstream and replace the original iocage in terms of use with this behaviour.

bartekrutkowski commented on iocell.8.txt:
```diff
@@ -1196,6 +1196,17 @@ PROPERTIES
+
+  Mounts and unmounts directories from the host into the jail root using
+  nullfs(5).
+  'mount_options' are passed direcly to the mount command and defaults to "rw"
```
I think it should default to 'ro' for security reasons, if not even fail, when no mode is provided.

nupfel replied:
i do not agree. i can tell you from experience using this a lot in production that you want 'rw' 99% of the time. in fact the only cases i wanted 'ro' so far was for single file volumes that are config and should not be changable from within the jail. lots of services will fail to even start if they can't write to a directory they need access to and won't tell you necessarily that it was a permission problem.

bartekrutkowski commented on iocell.8.txt:
```diff
@@ -1196,6 +1196,17 @@ PROPERTIES
+  nullfs(5).
+  'mount_options' are passed direcly to the mount command and defaults to "rw"
+  if ommitted.
+  If '/jail/dir' is also ommitted then the same value as '/host/dir' will be
```
I'd drop any configuration modes that can be done in more than one way, to avoid confusions and make automation easier, following Python's matra of 'there should be one and only one way to do it'.

nupfel replied:
you like python, others like perl. again, i do not agree and won't change this extremely convenient use of volumes because of your religion.

bartekrutkowski commented on lib/ioc-rc:
```diff
@@ -240,6 +240,10 @@ __start_jail () {
     __resolv_conf ${_fulluuid} ${_dataset} > ${_jail_path}/root/etc/resolv.conf
+
+    # mount optional volumes
+    __mount_volumes ${_fulluuid}
```
Missing quotation around _fulluuid?

nupfel replied:
yea, just like in the 3 lines above... looks consistent to me.

bartekrutkowski commented on lib/ioc-rc:
```diff
@@ -534,14 +538,15 @@ __stop_jail () {
     fi
 
     _name=$1
+    _no_unmount=$2
```
Missing quotation around $2?

nupfel replied:
but not around $1? can we stop this ridiculousness and keep going forward, please?

bartekrutkowski commented on lib/ioc-rc:
```diff
@@ -641,6 +646,11 @@ __stop_jail () {
 
+    # unmount optional volumes
+    if [ -z "${_no_unmount}" ] ; then
+        __unmount_volumes ${_fulluuid}
```
Missing quotation around _fulluuid?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+
+__print_exit () {
+    echo "$1"
+    __stop_jail $fulluuid no_unmount
```
Missing quotation around _fulluuid?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+# format of $volumes: /path/to/some/host/dir[:/location/in/jail/root[:mount,options]]
+# multiple volumes can be space separated
+__mount_volumes () {
+    local _uuid=$1
```
Missing quotation around $1?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+    local _mounts _mounted _retval
+    local volumes="$(__get_jail_prop volumes ${_uuid})"
+
+    if [ -z "${volumes}" -o "${volumes}" == "none" ] ; then
```
Given it is a multiconditional if, a comment what's happening here would be useful.

nupfel replied:
wow, yea. this is way too complicated for someone to read...

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+
+        # set default mount options if missing
+        if [ -z "${_mount_options}" ] ; then
+            _mount_options="rw"
```
Always deafult to least possible set of privileges.

nupfel replied:
see above

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+
+        # source path checking
+        if [ -z "${_source}" ] ; then
+            __print_exit "  ERROR: volume source cannot be empty: ${_source}"
```
Shouldn't __die be used here and in occurences below?

nupfel replied:
no, and if you read the code you would understand why. this also stops the jail before exiting. i renamed it `__stop_and_die` to be more obvious though.

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+        fi
+
+        # all paths must be absolute
+        if [ -z $(echo ${_source} | grep "^/") ] ; then
```
Missing quotation around _source?

nupfel replied:
fair enough

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+        if [ -z $(echo ${_source} | grep "^/") ] ; then
+            __print_exit "  ERROR: source volume path must be absolute: ${_source}"
+        fi
+        if [ -z $(echo ${_dest} | grep "^/") ] ; then
```
Missing quotation around _dest?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+                done
+            fi
+
+            __stop_jail ${_uuid} no_unmount
```
Missing quotation around _uuid?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+}
+
+__unmount_volumes () {
+    local _uuid=$1
```
Missing quotation around $1?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+            _source=${_volume}
+        else
+            # split into source, dest and mount options
+            _source=$(echo ${_volume} | cut -d: -f1)
```
Missing quotation around _volume?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+            rm "${_full_dest}"
+            _retval=$?
+        else
+            # not supported or nothing to do
```
If this is really how it should work, that else seems redundant, isn't it?

nupfel replied:
you did see the `continue` on the next line down, right? and you know that it instantly continues the for loop and therefor not executing the code further down? also i left this to maybe support other file types later if needed.

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+            continue
+        fi
+
+        if [ ${_retval} -eq 0 ] ; then
```
Missing quotation around _retval?

bartekrutkowski commented on lib/ioc-volumes:
```diff
@@ -0,0 +1,196 @@
+        if [ ${_retval} -eq 0 ] ; then
+            echo "    ${_source}"
+        else
+            echo "  ERROR: Unmounting volume ${_source} from ${tag}:${_dest} FAILED"
```
Shouldn't __die be used here?

bartekrutkowski commented:
@nupfel I've reviewed this pull request and I like the idea, very good first attempt that requires just a bit more work.

nupfel commented on doc/source/advanced-use.rst:
```diff
@@ -249,3 +249,30 @@ Now simply create a jail and supply the pkgs.txt file:
+
+Jails can have shared directories from the host which is implemented using nullfs(5). The idea is based on volumes used in the docker world. Files can also be Volumes and are copied over at jail startup.
+
+The ``volumes`` property has the following syntax:
```
no, it's volumes. did you read the documentation?

bartekrutkowski replied:
My point wasn't that this is not coherent with the documentation, but that it seems the pattern is to use singular forms instead of plural, for example 'ip4_addr' not 'ip4_addrs', 'source_path' not 'source_paths' and so on, and so forth.

nupfel replied:
that is just bad documentation style and i'm not going to continue the author's misleading style. it should be ip4_addrs really if it allows multiples.

rspeed replied:
Consistency seems more important to me. It would be better to use `volume`, then switch all properties with the same issue to plural forms separately and at the same time.

bartekrutkowski commented:
@nupfel Are you planning to update this PR with revised code?

rostwald commented:
Isn't this feature equivalent with just mounting or delegating additional ZFS-datasets in/to the jail? Thats what I'm doing when I need to share/deploy/move data between jails or need some "floating" storage. 

OTOH one might able to build other interesting things with nullfs-mounts that aren't possible with plain ZFS...

nupfel commented:
yes, i'm planning on updating this PR with some of the revised code at least, but I'm far too busy at the moment with real life problems and events, so i has to wait a bit, sorry.

bartekrutkowski commented:
@nupfel Any news on update for this PR?

nupfel commented:
i updated my branch with your requested changes where appropriate and squashed it down to single commit again. for some reason it does not seem to show in this PR though. so feel free to merge it yourself or let me know if you won't in which case i will submit my own port to finally get this code out as this has already been tested and used in production for a while now.

bartekrutkowski commented:
Rejecting, following the author's suggestion, given missing corrections requested.
